### PR TITLE
Formalize range allocation with a new data structure

### PR DESCRIPTION
**Pull request by mulkieran**
Supercedes #2361

### CODE REVIEW

**Comment by mulkieran** [comment]
Passed all tests.

File: src/engine/strat_engine/backstore/blockdev.rs [hunk 1]
```diff
@@ -144,8 +147,10 @@ impl StratBlockDev {
 
     /// Find some sector ranges that could be allocated. If more
     /// sectors are needed than are available, return partial results.
-    /// If all sectors are desired, use available() method to get all.
-    pub fn request_space(&mut self, size: Sectors) -> (Sectors, Vec<(Sectors, Sectors)>) {
+    /// If all available sectors are desired, don't use this function.
+    /// Write a simple request_all() function in the RangeAllocator
+    /// implmentation to get the result much more efficiently.
```
Comment (jbaublitz) [inline src/engine/strat_engine/backstore/blockdev.rs:152]: It looks like there's a typo.

File: src/engine/strat_engine/backstore/blockdev.rs [hunk 2]
```diff
@@ -144,8 +147,10 @@ impl StratBlockDev {
 
     /// Find some sector ranges that could be allocated. If more
     /// sectors are needed than are available, return partial results.
-    /// If all sectors are desired, use available() method to get all.
-    pub fn request_space(&mut self, size: Sectors) -> (Sectors, Vec<(Sectors, Sectors)>) {
+    /// If all available sectors are desired, don't use this function.
+    /// Write a simple request_all() function in the RangeAllocator
```
Comment (jbaublitz) [inline src/engine/strat_engine/backstore/blockdev.rs:151]: Is this a TODO?

File: src/engine/strat_engine/backstore/range_alloc.rs [hunk 3]
```diff
@@ -12,182 +15,386 @@ use crate::{
...
+        &self,
+        prev: Option<Sectors>,
+        next: Option<Sectors>,
+        range: &(Sectors, Sectors),
+    ) -> StratisResult<(
+        Option<(Sectors, Sectors)>,
+        (Sectors, Sectors),
+        Option<(Sectors, Sectors)>,
+    )> {
+        let (start, len) = range;
```
Comment (jbaublitz) [inline src/engine/strat_engine/backstore/range_alloc.rs:110]: Can pattern matching make this simpler? An example like:

```rust
fn main() {
    let tuple = &(1, 0);
    let &(rc, rs) = tuple;
    println!("{}", rc + 1);
}
```
seems to compile so `rc` is not a reference as long as `Copy` is implemented on the data type which it is for `Sectors`.

File: src/engine/strat_engine/backstore/range_alloc.rs [hunk 4]
```diff
@@ -12,182 +15,386 @@ use crate::{
...
+
+        Ok((lhs, res, rhs))
+    }
+
+    /// Insert specified range into self. Return an error if there is any
+    /// overlap between the specified range and existing ranges. If the range
+    /// is contiguous with an existing range, combine the two.
+    /// Inserting a 0 length range has no effect.
+    pub fn insert(&mut self, range: &(Sectors, Sectors)) -> StratisResult<()> {
+        let (start, len) = range;
```
Comment (jbaublitz) [inline src/engine/strat_engine/backstore/range_alloc.rs:190]: Similar question about pattern matching here.

**Review by jbaublitz — CHANGES_REQUESTED** [review]

File: src/engine/strat_engine/backstore/range_alloc.rs (hunk 3 above)
Comment (mulkieran) [reply src/engine/strat_engine/backstore/range_alloc.rs:110]: Yup, that's better!

**Comment by mulkieran** [comment]
Ok! This is ready again.

**Comment by mulkieran** [comment]
Test failure is stable clippy, ignore.

**Review by jbaublitz — APPROVED** [review]

**Pull request closed without merging**